### DISCUSSION TITLE
Improving Grafana Variables UI and features

### DISCUSSION THREAD

**Discussion by skeagy** [grafana/grafana#68753] [General]
I've been using grafana intensively for several months, currently with grafana-9.5.2-r0. Overall tool is amazing and I love it, but I very often find myself frustrated and wrestling against the limitations of grafana variables to achieve the dashboard functionality I want. That wrestling is where I spend the VAST majority of my time. Following is a list of various UI and functional capability improvements I'd like to see:

1. For promql data source variables, use the same wizard and code highlighting and whitespace handling as for the timeseries queries in a panel. There is a minor improvement in 9.5.2 to not simply fail a query because of newlines... but the newlines are still automatically deleted and the visualization improvement is lost whenever focus shifts away. Further, there is no color highlighting as in the panel query syntax that further helps a user understand the query. Why not leverage the same code for visual presentation of queries in panels or when defining dashboard variables?

2. The rules for when backslash escapes are needed are not the same between queries in a panel and queries in a variable definition. This gets pretty ugly when using backslashes for regex meaning or using literal backslashes as part of a regex string. I generally figure out what I need for each case, but it takes me several hours of banging my head before I figure it out, sometimes by putting things in a ${variable:raw} format adding an unnecessary layer of complexity.

3. For promql data sources, the regex field lets you define (?<text>()) and (?<value>()) capture groups. This can be creatively exploited but is very limiting for what I need. For example, I am overloading it with "host" and "interface" attributes in "text" and "value" fields right now, and that breaks when two different hosts have the same value... uniqueness of values appears to be enforced. It would be nice to have a checkbox to 'enforce uniqueness of 'value' contents, for cases like mine when it is essentially a hidden intermediate variable and not exposed to the dashboard user. The reason I link host and interface in a single variable is because I use it as a repeating row variable so I can build a series of charts that leverage the unique instance of host and interface for each row. This idea is powerful, but not fully realizable because of these implementation limitations.

4. A better implementation of this text/value and regex concept for variables would be to have a field representing the string for 'text' and another field representing string for 'value', and allow us to specify $2-$1etc. for more flexible formatting. This is critical for example to enable 3 or more labels (or sub-parts of labels) to be used as part of the derived text : value pair for each line.

5. An even better implementation of the overall concept of a 2-part variable is to have a multi-part variable. Why limit us to 'text' and 'value' capture groups on the regex part of variable definition? That seems like an organic outgrowth of the custom variable definition, but I have many use cases for having multiple variables keyed off of a single selected field. Why not extend my UI suggestion from  bullet point 4 above to be an extensible list of hashkey1, hashkey2, hashkey3, etc. with entries for the name of that hashkey and the field for regex substitution to populate the value of that hashkey. Then we can refer to the values of the different hashkeys like ${variable:mykey1}, ${variable:mykey2}, $variable:mykey3, etc.

Without having bullet point 5, I'm on the edge of creating an independent database just to populate lookup hashes of stuff so I can make a SQL query variable in grafana to lookup multiple variables based on a single key. That seems like a brittle add-on dependency and configuration complexity on my end as a grafana admin, when all I want to do is set 3 or more variables based on the value of one other. This has to be a super common need for anyone who goes beyond the most vanilla dashboard creation.

Thanks for listening :) Please let me know if there are better ways to do what I'm trying to achieve. I've pored over documentation and googling to figure this stuff out, and done some creative stretching beyond documentation to make stuff work, and looking to make the core grafana product better now.

**Comment by skeagy** [comment]
I've got hundreds of potential dashboard users within our network organization, visualizing metrics that span 50+ million unique label combinations with terrabytes of data. I've got that data in hand, doing some cool stuff with it already, but what I'm missing and asking for here is a HUGE impediment to doing what we want and building this into our workflow. It's a show-stopper feature for us adopting Grafana in a bigger way.

Here is a specific example for chaining custom variable...

A part of my PromQL (well, MetricsQL) query looks like this:
![promql_query_clause](https://github.com/grafana/grafana/assets/23060915/6f28ae72-ace8-4667-a619-f474b3a6fec7)



The problem is, the UI for the dashboard variables defining the business hours is a bit clunky:
![dashboard_variable_ui2](https://github.com/grafana/grafana/assets/23060915/97242ae8-adf3-4418-a8a4-dc22d520b83d)



Instead of asking users to select a start and stop hour, I'd like to have a set of policies in a drop-down list and have that drive the creation of the back-end promql that explicitly describes the policy.

I tried to do that with the biz_hrs variable, using the entire promql clause I have in the first image. Here is a sample of how I defined the biz_hrs variable of type Custom:
![desired_syntax_for_chaining_vars_within_custom_var](https://github.com/grafana/grafana/assets/23060915/cf1096de-2c90-4eed-bb77-45d7e7f21379)

But... this fails because Grafana doesn't expand the ${tz} variable! The whole clause with unprocessed variables is stuck into the promql that is sent as the query, and the promql server can't process it.

I can't see how to break apart the promql to leave the day of week and open and close times all contiguous as part of a single drop-down variable, while also letting the timezone be independently defined and inserted into the appropriate parts of this query.

Grafana dashboard variable expansion/chaining only works when querying an external data source, and if this was a conscious decision, it seems wrong to me. Why should I need to hack together an external MariaDB or SQLIte just to make a SQL query to fetch some hacked up  data rather than expanding/chaining the variable natively within a Custom variable? It should work just like external query variables. This specific feature limitation (no support for chaining variables in variables of type Custom) has blocked me from doing what I wanted many times.

If this were supported, then it would also give a great work-around for other problems I tried to work through with Grafana folks related to global label uniqueness, combining multiple labels to generate a globally unique label, etc.

It seems to be a small change with big benefits to unlocking dashboard developer creativity. The main unintended consequences I can foresee are the ugliness of defining escape characters (e.g. \$ or \$\{ to support literal versions of those characters) and how to escape the escape sequences. Those problems have been solved in sed, awk, perl... don't need to develop a new approach.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). We are migrating this discussion to an issue and closing the discussion. The issue is #82579. Feel free to continue the discussion around this there. Thank you!
